src/services/rickmorty.service.mjs

```diff
@@ -0,0 +1,76 @@
+import axios from "axios";
+
+const api = axios.create({
+  baseURL: "https://rickandmortyapi.com/graphql",
+});
+
+const MAX_PAGE = Math.ceil(671 / 20);
+const MAX_ITEMS_PER_PAGE = 20;
+
+function randomPage(maxPage) {
+  return Math.floor(Math.random() * maxPage + 1);
+}
+
+function buildQuery() {
+  const query = `
+  query {
+    characters(page: ${randomPage(MAX_PAGE)}) {
+      info {
+        count
+      }
+      results {
+        name,
+        image
+      }
+    }
+  }`;
+  return query;
+}
+
+function createCardObject(character, index) {
+  if (!character) throw new Error("Character is undefined");
+  return {
+    id: index + 1,
+    name: character.name,
+    image: character.image,
+  };
+}
+
+async function getCards(amount) {
+  try {
+    if (!amount || amount <= MAX_ITEMS_PER_PAGE) {
+      const query = buildQuery();
+      const res = await api.post("/", { query });
+      return res.data.data.characters.results
+        .map((char, index) => createCardObject(char, index))
+        .slice(0, amount);
+    } else {
+      const pageQueries = [];
+
+      // call multiple pages if the card amount is bigger than MAX_ITEMS_PER_PAGE
+      for (let i = 0; i < Math.ceil(amount / MAX_ITEMS_PER_PAGE); i++) {
+        const query = buildQuery();
+        pageQueries.push(api.post("/", { query }));
+      }
+
+      const cardPages = await Promise.all(pageQueries);
+
+      let cards = [];
+
+      cardPages.forEach((page) => {
+        cards = [
```

Comment:
Du iterierst für jede cardPage über alle bisherigen cardPages, d.h. du hast hier O(n²), könntest es aber in O(n) lösen. Vermutlich kein Bottleneck, so viel pages brauchen wir ja nicht und "everything is fast for small n" =D

Reply:
Wo iteriere ich für jede page über alle bisherigen cardpages? ich fülle das Cards array mit dem spread operator mit den neuen gemappten erstellen playcards. 

Man könnte aber das verbessern, weil ich über alle characters iteriere, obwohl ich evtl. nicht alle benötige.

Reply:
Nehmen wir an unsere Anzahl der pages hat Größenordnung `n` und beinhaltet jeweils `m` Elemente. Du callst in jeder Iteration von `forEach` `...cards`. Das erstell nen iterator und iteriert über diesen, bis er exhausted ist. Dabei wächst `cards` mit jeder iteration um `m` und das insgesamt `n` mal. insgesamt iterierst du also `n` mal über durchschnittlich (nach Gauß) `(m*n/2)` Elemente. Das wäre O(n) * O(m*n/2). m und 2 sind konstant und somit vernachlässigbar. Das heißt du iterierst in der Größenordnung von O(n) * O(n) = O(n²).